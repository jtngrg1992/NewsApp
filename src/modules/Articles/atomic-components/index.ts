export * from './Article.row';
export * from './NavigationButton';
