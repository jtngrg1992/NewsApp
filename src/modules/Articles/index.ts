export {default as ArticlesListingScreen} from './ArticlesList.screen';
