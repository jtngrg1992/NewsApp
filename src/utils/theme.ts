export default {
  primaryColor: '#48e4c2',
  secondaryColor: '#9a9a9a',
  backgroundColor: '#fafafa',
  onPrimary: '#fff',
  onBackground: '#000',
  onBackgroundLight: '#808080',
  onSecondary: '#fff',
};
