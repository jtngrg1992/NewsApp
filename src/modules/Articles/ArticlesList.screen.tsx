import React, {useCallback, useEffect, useLayoutEffect} from 'react';
import {useFocusEffect, useNavigation} from '@react-navigation/native';
import {
  Alert,
  FlatList,
  ListRenderItemInfo,
  RefreshControl,
  View,
} from 'react-native';
import {useDispatch} from 'react-redux';
import {
  useArticleList,
  useArticlesFetchError,
  useIsFetchingArticles,
} from '@news/hooks';
import {fetchArticles, Article} from '@news/model/articles';
import {ErrorTitle} from '@news/constants';
import {Images} from '@news/images';
import {ArticleRow, NavigationButton} from './atomic-components';

export default () => {
  const dispatch = useDispatch();
  const articles = useArticleList();
  const isFetching = useIsFetchingArticles();
  const error = useArticlesFetchError();
  const navigation = useNavigation();

  useFocusEffect(
    React.useCallback(() => {
      dispatch(fetchArticles());
    }, []),
  );

  useLayoutEffect(() => {
    navigation.setOptions({
      headerLeft: () => <NavigationButton image={Images.menu} />,
      headerRight: () => (
        <View style={{flexDirection: 'row', alignItems: 'center'}}>
          <NavigationButton image={Images.search} />
          <NavigationButton image={Images.more} />
        </View>
      ),
    });
  }, []);

  useEffect(() => {
    console.log(error);
    if (error) {
      Alert.alert(ErrorTitle, error.message);
    }
  }, [error]);

  const handleItemPress = useCallback((item: Article) => {
    navigation.navigate('Detail', {article: item});
  }, []);

  const renderItem = useCallback(
    ({item}: ListRenderItemInfo<Article>) => (
      <ArticleRow
        testID={`article-row-${item.id}`}
        article={item}
        onPress={() => handleItemPress(item)}
      />
    ),
    [],
  );

  return (
    <FlatList
      style={{flex: 1}}
      refreshControl={<RefreshControl refreshing={isFetching} />}
      refreshing={isFetching}
      data={articles}
      renderItem={renderItem}
      keyExtractor={(item) => item.id.toString()}
    />
  );
};
