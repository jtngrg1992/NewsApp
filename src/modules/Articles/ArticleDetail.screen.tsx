import {RouteProp} from '@react-navigation/native';
import React, {useMemo} from 'react';
import {View, StyleSheet, Image, Text, ScrollView} from 'react-native';
import {Images} from '../../images';
import {ArticleRoutes} from '../../routers/ArticlesRouter';
import theme from '../../theme';

type DetailScreenRouteProp = RouteProp<ArticleRoutes, 'Detail'>;

export default ({route}: {route: DetailScreenRouteProp}) => {
  const article = useMemo(() => route.params.article, [route?.params.article]);

  const {title, abstract, media, author, published} = article;

  return (
    <ScrollView style={styles.container} stickyHeaderIndices={[0]}>
      <View style={styles.imageContainer}>
        {media?.largeImage && (
          <Image
            testID="article-large-image"
            style={styles.articleImage}
            source={{uri: media.largeImage}}
            resizeMode="cover"
          />
        )}
      </View>
      <View style={styles.detailContainer}>
        <Text testID="article-title" style={styles.title}>
          {title}
        </Text>
        <View style={[styles.row, {marginTop: 10}]}>
          <View style={styles.spacer}>
            <Text testID="article-author" style={styles.infoText}>
              {author}
            </Text>
          </View>
          <View style={[styles.row, {flex: 1, justifyContent: 'flex-end'}]}>
            <Image style={styles.calendarIcon} source={Images.calendar} />
            <Text testID="article-date" style={styles.infoText}>
              {published}
            </Text>
          </View>
        </View>
        <Text testID="article-abstract" style={styles.abstract}>
          {abstract}
        </Text>
      </View>
    </ScrollView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.backgroundColor,
  },
  imageContainer: {
    height: 220,
    backgroundColor: theme.secondaryColor,
    width: '100%',
  },
  articleImage: {
    flex: 1,
  },
  detailContainer: {
    paddingHorizontal: 15,
    paddingVertical: 10,
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
    color: theme.onBackground,
  },
  abstract: {
    fontSize: 17,
    color: theme.onBackgroundLight,
    fontWeight: '500',
    marginTop: 15,
  },
  row: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  spacer: {
    flex: 1,
  },
  infoText: {
    fontSize: 14,
    color: theme.onBackgroundLight,
    flexWrap: 'wrap',
  },
  calendarIcon: {
    height: 20,
    width: 20,
    tintColor: theme.onBackgroundLight,
    marginRight: 10,
  },
});
