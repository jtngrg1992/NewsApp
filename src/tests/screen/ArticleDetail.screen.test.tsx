import {shallow, ShallowWrapper} from 'enzyme';
import React from 'react';
import {ArticleDetailScreen} from '../../modules/Articles';
import mockArticles from '../test-data/articles';
import {testID} from '../utils';

describe('unit/snapshot tests for ArticleDetailScreen', () => {
  let wrapper: ShallowWrapper;

  const mockArticle = mockArticles[0];

  beforeEach(() => {
    wrapper = shallow(
      <ArticleDetailScreen
        // @ts-ignore
        route={{
          params: {
            article: mockArticle,
          },
        }}
      />,
    );
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correct image in the header', () => {
    expect(
      wrapper.findWhere(testID('article-large-image')).prop('source'),
    ).toEqual({uri: mockArticle.media?.largeImage});
  });

  it('should render correct article title in designated `Text` element', () => {
    expect(wrapper.findWhere(testID('article-title')).prop('children')).toBe(
      mockArticle.title,
    );
  });

  it('should render correct author and publishing date in designated `Text` elements', () => {
    expect(wrapper.findWhere(testID('article-author')).prop('children')).toBe(
      mockArticle.author,
    );
    expect(wrapper.findWhere(testID('article-date')).prop('children')).toBe(
      mockArticle.published,
    );
  });

  it('should render correct article abstract in designated `Text` element', () => {
    expect(wrapper.findWhere(testID('article-abstract')).prop('children')).toBe(
      mockArticle.abstract,
    );
  });

  it('should not render article image if not present', () => {
    const mockArticle = mockArticles[1];
    wrapper = shallow(
      <ArticleDetailScreen
        // @ts-ignore
        route={{
          params: {
            article: mockArticle,
          },
        }}
      />,
    );
    expect(wrapper.find(testID('article-large-image')).exists()).toBeFalsy();
  });
});
