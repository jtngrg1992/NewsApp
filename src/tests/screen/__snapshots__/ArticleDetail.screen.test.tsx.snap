// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit/snapshot tests for ArticleDetailScreen should render correctly 1`] = `
<ScrollView
  stickyHeaderIndices={
    Array [
      0,
    ]
  }
  style={
    Object {
      "backgroundColor": "#fafafa",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#9a9a9a",
        "height": 220,
        "width": "100%",
      }
    }
  >
    <Image
      resizeMode="cover"
      source={
        Object {
          "uri": "some.other.url",
        }
      }
      style={
        Object {
          "flex": 1,
        }
      }
      testID="article-large-image"
    />
  </View>
  <View
    style={
      Object {
        "paddingHorizontal": 15,
        "paddingVertical": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#000",
          "fontSize": 20,
          "fontWeight": "bold",
        }
      }
      testID="article-title"
    >
      Mock title
    </Text>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          },
          Object {
            "marginTop": 10,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#808080",
              "flexWrap": "wrap",
              "fontSize": 14,
            }
          }
          testID="article-author"
        >
          Mock Author
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            },
            Object {
              "flex": 1,
              "justifyContent": "flex-end",
            },
          ]
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/images/calendar.png",
            }
          }
          style={
            Object {
              "height": 20,
              "marginRight": 10,
              "tintColor": "#808080",
              "width": 20,
            }
          }
        />
        <Text
          style={
            Object {
              "color": "#808080",
              "flexWrap": "wrap",
              "fontSize": 14,
            }
          }
          testID="article-date"
        >
          2021-01-05
        </Text>
      </View>
    </View>
    <Text
      style={
        Object {
          "color": "#808080",
          "fontSize": 17,
          "fontWeight": "500",
          "marginTop": 15,
        }
      }
      testID="article-abstract"
    >
      Mock Abstract
    </Text>
  </View>
</ScrollView>
`;
